etcdserver/server.go

```diff
@@ -468,106 +468,86 @@ func (s *EtcdServer) ReportSnapshot(id uint64, status raft.SnapshotStatus) {
 	s.r.ReportSnapshot(id, status)
 }
 
+type raftProgress struct {
+	confState raftpb.ConfState
+	snapi     uint64
+	appliedi  uint64
+}
+
+// newApplier buffers apply packets on apply input channel and a
+// raftProgress output channel. This is so raftNode won't block on
+// sending new applies, timing out (since applies can be slow). Close
+// the apply channel to cancel the buffer's goroutine and read on
+// raftprogc until closed so pending operations clear.
+func (s *EtcdServer) newApplier(rp raftProgress) (chan<- apply, <-chan raftProgress) {
+	applyc := make(chan apply)
+	raftprogc := make(chan raftProgress)
+	go func() {
+		defer close(raftprogc)
+		pending := []apply{}
+		donec := make(chan struct{})
+		for applyc != nil || len(pending) > 0 {
+			// serialized function
+			f := func(a apply) {
+				s.applyAll(&rp, &a)
+				raftprogc <- rp
+				donec <- struct{}{}
+			}
+			select {
+			// launch if no pending apply packet, queue up the rest
+			case apply, ok := <-applyc:
+				if !ok {
+					applyc = nil
+					break
+				}
+				pending = append(pending, apply)
+				if len(pending) == 1 {
+					go f(pending[0])
```

Comment:
probably we can add a TODO here: Create a long running worker routine to avoid routine creation on normal non-waiting path.